Add meta$ parsing

dieghernan opened the pull request:
Closes #7 

I have adapted the chunk of code of **{cffr}** for parsing `meta$` tags on CITATION (it has been extensively tested on ~2000 packages: see [here](https://github.com/dieghernan/cffr/blob/main/tests/testthat/test_local_installation/_snaps/full_cff.md)) to **{codemeta}**.

Other changes you would see on the PR:
* Test suit added with snapshots. Also new test files added, included `CITATION` from  `rmardown`.
* Legacy utils functions of **{codemetar}** deleted. They were not used in **{codemeta}**.
* A safer way to `grep` the `source_code_domains`. Previous version thrown sometimes an error.
* RoxygenNote updated due to new version on CRAN of **{roxygen2}**.
* **{testthat}** upgraded, since snapshots were implemented on `(>= 3.0.0)`. This also means that `expect_is`  (deprecated on `(>= 3.0.0)` should be replaced with `expect_type`.


Ping @dpprdan 

Regards

dieghernan committed:
Add meta$ parsing

dieghernan committed:
No more "\n" on description field

dieghernan commented:
I needed to perform an extra change, now it is ready

dpprdan commented on R/parse_citation.R:
```diff
@@ -131,27 +131,25 @@ guess_citation <- function(path) {
 
   # Read and parse CITATION
-  bib <- read_citation_with_encoding(citation_path, description$get("Encoding"))
+  bib <- read_citation_with_encoding(citation_path, meta)
```
```suggestion
  bib <- utils::readCitationFile(citation_path, meta)
```

I am reasonably sure you can get rid of `read_citation_with_encoding()` altogether, because `readCitationFile()` should not error anymore with all meta info present, i.e. `citation(auto = meta)` should work and the error, `read_citation_with_encoding()` is trying to catch is never triggered. (I'm reasonably sure, because `readCitationFile(citfile, meta)` is what [`utils::citation()` does](readCitationFile(citfile, meta))).

There may be some use for the "get rid of the `citation(auto = meta) line" approach, though, because the CodeMeta citation field is supposed to (only) contain references to other works, not to the source code itself, IIUC.

> A citation or reference to _another_ creative work, such as another publication, web page, scholarly article, etc.
Source: https://codemeta.github.io/terms/ under "citation", emphasis mine.

So `read_citation()` could remove `citation(auto = meta)` if it is present. I wonder if this is the most efficient way to get rid of a self-citation though.

dieghernan replied:
See https://github.com/cboettig/codemeta/pull/9/commits/6aca0c78872c92e459c270908076ae6b71d26cd8

dieghernan committed:
Update R/parse_citation.R

Co-authored-by: Daniel Possenriede <possenriede@gmail.com>

dieghernan committed:
Remove `read_citation_with_encoding`

And refactor tests, thanks @dpprdan

See
https://github.com/cboettig/codemeta/pull/9#discussion_r708647950

dieghernan commented:
GHA is erroring for reasons not related with the PR. Locally I run `rcmdcheck::rcmdcheck(args = c("--as-cran"))` and 
`devtools::check()` succesfully: 

**0 errors √ | 0 warnings √ | 0 notes √**

GHA log (<https://github.com/cboettig/codemeta/runs/3607055303?check_suite_focus=true>):

```

2021-09-15T06:48:54.5805180Z ##[group]Run remotes::install_deps(dependencies = TRUE)
2021-09-15T06:48:54.5805990Z [36;1mremotes::install_deps(dependencies = TRUE)[0m
2021-09-15T06:48:54.5806610Z [36;1mremotes::install_cran("rcmdcheck")[0m
2021-09-15T06:48:54.5896480Z shell: /usr/local/bin/Rscript {0}
2021-09-15T06:48:54.5896890Z env:
2021-09-15T06:48:54.5897520Z   R_REMOTES_NO_ERRORS_FROM_WARNINGS: false
2021-09-15T06:48:54.5897980Z   RSPM: 
2021-09-15T06:48:54.5899590Z   GITHUB_PAT: ***
2021-09-15T06:48:54.5900140Z   R_LIBS_USER: /Users/runner/work/_temp/Library
2021-09-15T06:48:54.5900610Z   TZ: UTC
2021-09-15T06:48:54.5901010Z   _R_CHECK_SYSTEM_CLOCK_: FALSE
2021-09-15T06:48:54.5901530Z   NOT_CRAN: true
2021-09-15T06:48:54.5901910Z ##[endgroup]
2021-09-15T06:48:56.4146950Z waldo     (0.3.0 -> 0.3.1) [CRAN]
2021-09-15T06:48:56.4155630Z jquerylib (NA    -> 0.1.4) [CRAN]
2021-09-15T06:48:56.4173260Z rmarkdown (2.10  -> 2.11 ) [CRAN]
2021-09-15T06:48:56.4184520Z Installing 3 packages: waldo, jquerylib, rmarkdown
2021-09-15T06:48:56.4412540Z Installing packages into ‘/Users/runner/work/_temp/Library’
2021-09-15T06:48:56.4451360Z (as ‘lib’ is unspecified)
2021-09-15T06:48:56.6147170Z ##[error]Error: (converted from warning) unable to access index for repository /src/contrib:
2021-09-15T06:48:56.6150920Z   cannot open URL '/src/contrib/PACKAGES'
2021-09-15T06:48:56.6152870Z Execution halted

```

cboettig commented:
fixed :-)
